libcob/profiling.c

```diff
@@ -0,0 +1,464 @@
+{
+	cobglobptr = lptr;
+        cobsetptr  = sptr;
+}
+
+struct cob_prof_module *
+cob_prof_init_module (cob_module *module,
+		      struct cob_prof_procedure *procedures,
+		      size_t procedure_count)
+{
+	prof_init_static();
+	if (is_active){
+		struct cob_prof_module *info;
+		struct cob_prof_module_list *item;
+
+		info = cob_malloc (sizeof(struct cob_prof_module));
+		info->total_times = cob_malloc ( procedure_count * sizeof(cob_ns_time) );
+		info->called_count = cob_malloc ( procedure_count * sizeof(unsigned int) );
+		info->procedure_recursions = cob_malloc ( procedure_count * sizeof(unsigned int) );
+		info->procedures = procedures;
+		info->procedure_count = procedure_count;
+
+		item = cob_malloc (sizeof(struct cob_prof_module_list));
+		item->info = info;
+		item->next = prof_info_list;
+		prof_info_list = item;
+		return info;
+	}
+	return NULL;
+}
+
+static void
+cob_prof_realloc_arrays (void)
+{
+	int new_size = max_prof_depth * 2 + 16;
+
+	if (new_size > cobsetptr->cob_prof_max_depth)
+		new_size = cobsetptr->cob_prof_max_depth;
+
+	if (max_prof_depth >= new_size){
+		int i;
+		fprintf (stderr, "[cob_prof] Profiling overflow at %d calls, aborting profiling.\n", current_idx);
+		fprintf (stderr, "  Last 10 calls on stack:\n");
+		for (i=0;i<10;i++){
+			struct cob_prof_module *info = called_runtimes[current_idx-1-i];
+			int proc_idx = called_procedures[current_idx-1-i];
+			struct cob_prof_procedure *proc = info->procedures + proc_idx;
+			fprintf (stderr, "  * %s at %s:%d\n", proc->text,
+				 proc->file, proc->line);
+		}
+		is_active = 0;
+		return;
+	}
+
+	if (max_prof_depth){
+		start_times = cob_realloc (
+			start_times,
+			max_prof_depth * sizeof(cob_ns_time),
+			new_size * sizeof(cob_ns_time)
+			);
+		called_procedures = cob_realloc (
+			called_procedures,
+			max_prof_depth * sizeof(int),
+			new_size * sizeof(int)
+			);
+		called_runtimes = cob_realloc (
+			called_runtimes,
+			max_prof_depth * sizeof(struct cob_prof_module*),
+			new_size * sizeof(struct cob_prof_module*)
+			);
+
+	} else {
+		start_times = cob_malloc (new_size * sizeof(cob_ns_time));
+		called_procedures = cob_malloc (new_size * sizeof(int));
+		called_runtimes = cob_malloc (new_size * sizeof(struct cob_prof_module*));
+	}
+	max_prof_depth = new_size;
+}
+
+static void
+cob_prof_enter_procedure (struct cob_prof_module *info, int proc_idx)
+{
+	cob_ns_time t;
+
+	if (!is_active || info == NULL) return;
+
+	t = get_ns_time ();
+
+	current_idx++;
+	if (current_idx >= max_prof_depth) {
+		cob_prof_realloc_arrays();
+		if (!is_active) return;
+	}
+
+	called_procedures[current_idx] = proc_idx;
+	called_runtimes[current_idx] = info;
+	start_times[current_idx] = t;
+
+	info->procedure_recursions[proc_idx] ++;
+	info->called_count[proc_idx] ++;
+}
+
+static void
+cob_prof_exit_procedure (struct cob_prof_module *info, int proc_idx)
+{
+	/* Exit all the sections/paragraphs */
+	cob_ns_time t;
+
+	if (!is_active || info == NULL) return;
```

Comment:
Here and below: how can we get to `info == NULL`?

Reply:
Indeed, it's not supposed to happen, ever. But I like to have stronger checks in case the code evolves in the future and it becomes possible...